synapse/lib/stormlib/cache.py

```diff
@@ -0,0 +1,157 @@
+        except s_exc.BadSyntax as e:
+            raise s_exc.BadArg(mesg=f'Invalid callback query: {e.errinfo.get("mesg")}')
+
+        if not query.hasAstClass(s_ast.Return):
+            raise s_exc.BadArg(mesg='Callback query must return a value')
+
+        return FixedCache(self.runt, query, size=size)
+
+@s_stormtypes.registry.registerType
+class FixedCache(s_stormtypes.StormType):
+    '''
+    A StormLib API instance of a Storm Fixed Cache.
+    '''
+    _storm_locals = (
+        {'name': 'get', 'desc': 'Get an item from the cache by key.',
+         'type': {'type': 'function', '_funcname': '_methGet',
+                  'args': (
+                      {'name': 'key', 'type': 'any', 'desc': 'The key to lookup.'},
+                  ),
+                  'returns': {'type': 'any',
+                              'desc': 'The value from the cache, or the callback query if it does not exist', }}},
+        {'name': 'pop', 'desc': 'Pop an item from the cache.',
+         'type': {'type': 'function', '_funcname': '_methPop',
+                  'args': (
+                      {'name': 'key', 'type': 'any', 'desc': 'The key to pop.'},
+                  ),
+                  'returns': {'type': 'any',
+                              'desc': 'The value from the cache, or $lib.null if it does not exist', }}},
+        {'name': 'put', 'desc': 'Put an item into the cache.',
+         'type': {'type': 'function', '_funcname': '_methPut',
+                  'args': (
+                      {'name': 'key', 'type': 'any', 'desc': 'The key put in the cache.'},
+                      {'name': 'value', 'type': 'any', 'desc': 'The value to assign to the key.'},
+                  ),
+                  'returns': {'type': 'null', }}},
+        {'name': 'clear', 'desc': 'Clear all items from the cache.',
+         'type': {'type': 'function', '_funcname': '_methClear',
+                  'returns': {'type': 'null', }}},
+    )
+    _storm_typename = 'cache:fixed'
+    _ismutable = False
+
+    def __init__(self, runt, query, size=CACHE_SIZE_DEFAULT):
+        s_stormtypes.StormType.__init__(self)
+        self.runt = runt
+        self.size = size
+        self.query = query
+        self.locls.update(self.getObjLocals())
+
+        self.cache = s_cache.FixedCache(self._runCallback, size=size)
+
+    def __len__(self):
+        return len(self.cache)
+
+    def getObjLocals(self):
+        return {
+            'pop': self._methPop,
+            'put': self._methPut,
+            'get': self._methGet,
+            'clear': self._methClear,
+        }
+
+    async def _runCallback(self, key):
+        opts = {'vars': {'cache_key': key}}
+        async with self.runt.getSubRuntime(self.query, opts=opts) as subr:
+            try:
+                async for _ in subr.execute():
+                    await asyncio.sleep(0)
+            except s_stormctrl.StormReturn as e:
+                return await s_stormtypes.toprim(e.item)
```

Comment:
Current behavior is to default to `None` if no `StormReturn` or exception occurs. 1) Is that be behavior we want? 2) Can we add a test to capture what the intended behavior is?

Reply:
The fact that this function can leak StormCtrlFlow exceptions is its own problem. We probably want to catch all of those so we don't a function call doing unintended control flow things. Here is an example:

```
$c = $lib.cache.fixed( ${ if ($cache_key < (4)) { return (`HAHA {$cache_key} HEHE`) } else { break } } )

for $i in $lib.range((20)) {
    $v = $c.get($i)
    $lib.print(`{$i} -> {$v}`)
}

$lib.print(`Last value of was i={$i}`)
```

Produces the following:

```
0 -> HAHA 0 HEHE
1 -> HAHA 1 HEHE
2 -> HAHA 2 HEHE
3 -> HAHA 3 HEHE
Last value of was i=4
```

Reply:
The storm query already must contain a return statement, so I think it's clearer if we return None if that return() is not hit.  I clarified the docs.

Reply:
I'm not following your "leak StormCtrlFlow" example.  What is the intent of the user with that break?  Currently the behavior you're showing is the same as if you called a function in that way, so I think it makes sense to raise it instead of swallowing it, but happy to talk about it more.  I added a couple more tests for this type of behavior.

Reply:
The example is intended to show the intersection of throwing a exception ( StormCtrlFlow ) when making a function call on a Storm object [ the cache ] that cannot be caught with ``try ... catch ...`` constructs. I'm not certain we have many other cases of that being possible, outside of the `$lib.storm.eval` ?